### ISSUE TITLE
Can otterscan get the number of holders of a token?

### ISSUE THREAD

**Issue by TangMonk** [wmitsuda/otterscan#718]
Get the number of holders of a token and holders list, is it possible?

**Comment by wmitsuda** [comment]
Not for now, but I'm working on some indexes to do exactly what you are asking for. So, that'll be possible soon, but I'll required some extra disk space/sync time.

**Comment by wmitsuda** [comment]
will leave this open to reference it when this feature gets released.

**Comment by TangMonk** [comment]
I am planning to implement the functionality of Holders by myself. Should I traverse all the Transfers of a certain Token whenever there is a new Transfer Event and then calculate all the Holders data?

```mermaid
graph TD;
    A[A new Transfer Event occurs] --> B[Traverse all Transfer records of the Token];
    B --> C[Calculate Token balance for each address(balance > 0)];
    C --> D[Sort addresses by balance to get Holders list];

```

**Comment by wmitsuda** [comment]
yes, it's very resource intensive. I'm almost releasing the first indexer prototype, but need to optimize the holder calculation, it's taking over 10-11 hours in my machine, I think it can be merged with token transfers, which also traverse logs and make both take ~10 hours.

I'll do this test this week after finishing some tasks from the frontend backlog which are accumulating :)

**Comment by wmitsuda** [comment]
just for the record, looping through all token transfers on the fly doesn't scale well because (at least on erigon2) it requires reading logs from the disk, each log is raw serialized.

the indexer model I'm proposing will consist of a table holder -> token addresses, meaning: from the infinite universe of existing tokens, which ones a certain holder has ever touched, meaning it may have some balance.

then, for that small set, calculates the balance on the fly (tracking the balance in the DB is not reliable because of exotic tokens; balance() is a function, not data).

**Comment by dreadedhamish** [comment]
> the indexer model I'm proposing will consist of a table holder -> token addresses, meaning: from the infinite universe of existing tokens, which ones a certain holder has ever touched, meaning it may have some balance.

I'm still figuring out all the moving parts - when you way you are proposing an indexer model, is that a proposal you are making to Erigon to implement?

**Comment by wmitsuda** [comment]
yes, that's it.

just giving you a heads up, I already implemented that for erigon 2. what the original issue asked is the inverse: giving a token, what are the holders, and I think it shouuld not be hard to add this one.

I just didn't do it yet because I'm focusing in migrating everything done to erigon 3 before adding more indexes.

**Comment by dreadedhamish** [comment]
> yes, that's it.
> 
> just giving you a heads up, I already implemented that for erigon 2. what the original issue asked is the inverse: giving a token, what are the holders, and I think it shouuld not be hard to add this one.
> 
> I just didn't do it yet because I'm focusing in migrating everything done to erigon 3 before adding more indexes.

Awesome. I've been thinking through what Otterscan needs for feature-parity with blockscout or etherscan, and holders per token and tokens per holder are the big obvious ones.

**Comment by wmitsuda** [comment]
I'm pretty sure once we survive this erigon 2 -> 3 migration phase and we master the know-how of doing indexes in the new architecture, we'll be able to implement all missing indexes pretty easily/quikcly.
